### PR TITLE
[nl] Add date range support for highlight, bar, map, ranking, and scatter

### PR DESCRIPTION
**Pull request by chejennifer**
For charts that can only show a single date, when a date range is in the query, show the latest date within the date range.

Changes made:
- When there is an end date, during existence checks, also find the latest valid date for each place key and save it to exist_checks in the state
- When generating ChartSpec in the various handlers, pass in the latest dates from the state & then use that when building the charts
- For filter_with_single_var and filter_with_dual_var handlers, update them to use date when filtering and ranking children
- For filter_with_dual_var and time_delta_across_places, update them to get latest date from the additional existence checks done in those handlers
- update date ranges of the type "last x years" to only set a start date because technically it's start date until latest

example: female population in New York before 2019
prod: https://screenshot.googleplex.com/8o7Qyxj9PKB3cM2
with this change: https://screenshot.googleplex.com/4PNKxXxvdWsJNaj

### CODE REVIEW

**Comment by pradh** [comment]
I'll take a careful look soon...

File: server/lib/nl/common/utils.py
Comment (pradh) [inline server/lib/nl/common/utils.py:266]: use helper func?

File: server/lib/nl/fulfillment/types.py
Comment (pradh) [inline server/lib/nl/fulfillment/types.py:91]: nit: Used only if single-date/date-range is set?  document?

File: server/lib/nl/fulfillment/types.py
Comment (pradh) [inline server/lib/nl/fulfillment/types.py:149]: nit: used when?

File: server/lib/nl/fulfillment/types.py
Comment (pradh) [inline server/lib/nl/fulfillment/types.py:169]: ditto

File: server/lib/nl/fulfillment/time_delta_across_places.py
Comment (pradh) [inline server/lib/nl/fulfillment/time_delta_across_places.py:142]: hmm, should this happen only when `state.date_range` is set?

File: server/lib/nl/fulfillment/simple.py
Comment (pradh) [inline server/lib/nl/fulfillment/simple.py:96]: hmm, why should we call this always?

**Review by pradh — COMMENTED** [review]
Sorry for the late response.

I feel a bit uncomfortable at how intrusive the changes are, my expectation was that this should affect things when date_range is set, so I wonder if we could make that clearer in the code (left some comments)?

Also, should we add some focussed e2e tests?

File: server/lib/explore/existence.py
Comment (chejennifer) [reply server/lib/explore/existence.py]: done

File: server/lib/nl/common/rank_utils.py
Comment (chejennifer) [reply server/lib/nl/common/rank_utils.py]: done

File: server/lib/nl/common/rank_utils.py
Comment (chejennifer) [reply server/lib/nl/common/rank_utils.py:155]: done

File: server/lib/nl/common/rank_utils.py
Comment (chejennifer) [reply server/lib/nl/common/rank_utils.py:412]: The callers of this function may pass in empty string (where date doesn't matter) & it's this api specifically where 'LATEST' has to be set

File: server/lib/nl/common/utils.py
Comment (chejennifer) [reply server/lib/nl/common/utils.py:276]: done

File: server/lib/nl/fulfillment/types.py
Comment (chejennifer) [reply server/lib/nl/fulfillment/types.py:92]: done

File: server/lib/nl/fulfillment/types.py
Comment (chejennifer) [reply server/lib/nl/fulfillment/types.py:150]: done

File: server/lib/nl/fulfillment/types.py
Comment (chejennifer) [reply server/lib/nl/fulfillment/types.py:171]: done

File: server/lib/nl/fulfillment/time_delta_across_places.py
Comment (chejennifer) [reply server/lib/nl/fulfillment/time_delta_across_places.py:145]: yes, if state.date_range is not set, end_date will be empty

File: server/lib/nl/fulfillment/simple.py
Comment (chejennifer) [reply server/lib/nl/fulfillment/simple.py:98]: we only set latest_valid_date in exist_checks if there is a date_range, so this function will map all sv and places to empty string if there was no date range. Added comments in `get_sv_place_latest_date` and `ExistInfo`

**Comment by chejennifer** [comment]
> Sorry for the late response.
> 
> I feel a bit uncomfortable at how intrusive the changes are, my expectation was that this should affect things when date_range is set, so I wonder if we could make that clearer in the code (left some comments)?
> 
> Also, should we add some focussed e2e tests?

Added some more queries so that we cover date range in all the different single date chart types: bar, highlight, map, ranking, and scatter

File: server/lib/nl/fulfillment/time_delta_across_vars.py
Comment (pradh) [inline server/lib/nl/fulfillment/time_delta_across_vars.py:104]: It slightly feels to me that the data-structure for this and get_sv_place_facet_ids could be unified, since they both come into play when there's date range, right?  Would that make sense?

I'm also wondering why this is not conditioned under date_range check?

Comment (pradh) [reply server/lib/nl/fulfillment/time_delta_across_vars.py:104]: Feel free to defer to post submission :)

**Review by pradh — APPROVED** [review]
Sorry for the long delay.

Approving to unblock concurrent changes.  I'm not fully comfortable with this change, but will think more about refactoring (left a couple of follow up questions)...

File: server/lib/nl/common/rank_utils.py
Comment (chejennifer) [reply server/lib/nl/common/rank_utils.py:412]: done

File: server/lib/nl/fulfillment/time_delta_across_vars.py
Comment (chejennifer) [reply server/lib/nl/fulfillment/time_delta_across_vars.py:104]: actually, facet_ids is needed for timeline charts (both single date and date range) and place_latest_date is needed for single place charts like map and highlight and used for date range

& this is not conditioned under date_range check because get_sv_place_latest_date actually will return empty when there's no date range and there's a comment on the function explaining this

File: server/integration_tests/explore_test.py
Comment (chejennifer) [reply server/integration_tests/explore_test.py]: done

File: server/lib/nl/common/utils.py
Comment (chejennifer) [reply server/lib/nl/common/utils.py:303]: done

File: server/lib/nl/fulfillment/existence.py
Comment (chejennifer) [reply server/lib/nl/fulfillment/existence.py:96]: done

**Comment by chejennifer** [comment]
thanks for the review!

**Pull request merged**
